assets/src/hooks/useDetour.ts

```diff
@@ -27,37 +28,76 @@ export enum DetourState {
   Finished,
 }
 
-export const useDetour = ({ routePatternId, shape }: OriginalRoute) => {
-  const [snapshot, send] = useMachine(createDetourMachine)
+const selectFinishedState = (
+  snapshot: SnapshotFrom<typeof createDetourMachine>
+) => {
+  /*
+   * There's probably a better way to do this? Tags or maybe context?
+   * Tags seem more appropriate, but weird to manage Out-Of-Bounds state via tags.
+   * Maybe this entire API call could be moved to and managed by an actor
+   * combined with parallel child states within the state machine?
+   * -- https://stately.ai/docs/promise-actors */
+  const isFinishedDrawing = snapshot.matches({
+    "Detour Drawing": { Editing: "Finished Drawing" },
+  })
+  const isSharingDetour = snapshot.matches({ "Detour Drawing": "Share Detour" })
+  const isInFinishedDetourState = isFinishedDrawing || isSharingDetour
+  return isInFinishedDetourState
+}
+
+/** This selects the detour waypoints in-between the detour start and end points  */
+const selectWaypoints = (snapshot: SnapshotFrom<typeof createDetourMachine>) =>
+  snapshot.matches({
+    "Detour Drawing": { Editing: "Finished Drawing" },
+  })
+    ? snapshot.context.waypoints.slice(1, -1)
+    : snapshot.matches({ "Detour Drawing": { Editing: "Place Waypoint" } })
+    ? snapshot.context.waypoints.slice(1)
+    : []
+
+const selectStartPoint = (snapshot: SnapshotFrom<typeof createDetourMachine>) =>
+  snapshot.context.waypoints.at(0)
 
-  const [startPoint, setStartPoint] = useState<ShapePoint | null>(null)
-  const [endPoint, setEndPoint] = useState<ShapePoint | null>(null)
-  const [waypoints, setWaypoints] = useState<ShapePoint[]>([])
+const selectEndPoint = (snapshot: SnapshotFrom<typeof createDetourMachine>) =>
+  (selectFinishedState(snapshot) && snapshot.context.waypoints.at(-1)) ||
+  undefined
+
+export const useDetour = ({ routePatternId, shape }: OriginalRoute) => {
+  const [snapshot, send, createDetourActor] = useMachine(createDetourMachine)
+  const startPoint = useSelector(createDetourActor, selectStartPoint)
+  const endPoint = useSelector(createDetourActor, selectEndPoint)
+  /** The detour waypoints in-between the start and end point */
+  const waypoints = useSelector(createDetourActor, selectWaypoints)
+  const allPoints = snapshot.context.waypoints
+
+  const isInFinishedDetourState = useSelector(
+    createDetourActor,
+    selectFinishedState
+  )
 
   const { result: finishedDetour } = useApiCall({
     apiCall: useCallback(async () => {
-      if (startPoint && endPoint) {
+      /* Until we have "typegen" in XState,
+       * we need to validate these exist for typescript
+       *
+       * > [Warning] XState Typegen does not fully support XState v5 yet. However,
+       * > strongly-typed machines can still be achieved without Typegen.
+       * > -- https://stately.ai/docs/migration#use-typestypegen-instead-of-tstypes
+       */
+      if (isInFinishedDetourState && startPoint && endPoint) {
```

Comment:
Is it possible for `isInFinishedDetourState` to be `false` if `endPoint` isn't set?

Reply:
As a "fun" "game", I removed the check for `isInFinishedDetourState &&` and all the tests still passed. I think that also means that the `isInFinishedDetourState` can be removed.

I _think_ if you change how `startPoint`/`endPoint`/`waypoints` are stored per my comment above, that could make `selectFinishedState` unnecessary.

Reply:
Still thinking on this, but I think this counts as derived state, but that doesn't answer if the `startPoint` and `endPoint` should be stored vs `waypoints`, https://github.com/statecharts/statecharts/discussions/36

Reply:
https://github.com/mbta/skate/pull/2629